Add std.range.Option

JakobOvrum opened the pull request:
Modeled after Scala's [scala.Option](http://www.scala-lang.org/api/2.8.1/scala/Option.html) type. It's been discussed on the forums before, I found at least [this thread](http://forum.dlang.org/post/l87ivq$263r$1@digitalmars.com).

~~(I'll move it to std.experimental.typecons once #2945 is merged, as it establishes the module)~~

JakobOvrum commented:
I also have these:
```d
Option!T left(T, U)(auto ref Algebraic!(T, U) algebraic)
{
	if(auto p = algebraic.peek!T)
		return some(*p);
	else
		return option(none);
}

Option!T right(T, U)(auto ref Algebraic!(T, U) algebraic)
{
	if(auto p = algebraic.peek!U)
		return some(*p);
	else
		return option(none);
}
```
Not sure if I should include them in this PR.

JackStouffer commented on std/typecons.d:
```diff
@@ -6896,3 +6896,418 @@ unittest
     );
 }
+
+private enum isNullable(T) = is(typeof(T.init is null));
```
Could these be moved to std.traits? If they are helpful here then the chances of them being helpful to someone else is high.

JakobOvrum replied:
Yes, but it's outside the scope of this PR. They can be moved later. (I think std.allocator also has its own `isCopyable` template)

JakobOvrum commented:
Fixed linker error and added Option.orElse.

JakobOvrum commented:
`Option!T` now works for const and immutable types `T` (including as a range), with the limitation that it's only reassignable to `none`, it cannot be assigned to another `Option!T`.

`const Option!T` and `immutable Option!T` are usable to some extent as well, but they're not ranges as ranges need to be poppable.

JackStouffer commented on std/typecons.d:
```diff
@@ -6896,3 +6897,503 @@ unittest
+    assert(maybeValues.joiner.map!(std.uni.toUpper).joiner(" ").equal("HELLO WORLD"));
+}
+
+/// Use `foreach` or $(REF each, std, algorithm, iteration) with `joiner` to perform an operation with side effects on `Option`s:
```
> `foreach`

Why not use `$(D foreach)`. I thought that that was the standard. Same applies everywhere else

> $(REF each, std, algorithm, iteration)

this didn't link for some reason

JakobOvrum replied:
``` `…` ``` is the same as `$(D …)`.

`REF` is from D-Programming-Language/dlang.org/pull/1184.

JackStouffer commented on std/typecons.d:
```diff
@@ -6896,3 +6897,503 @@ unittest
+    assert(opt.assumeNonEmpty == b);
+}
+
+/// Use $(LREF _Option.or) to convert to $(REF Algebraic, std, variant):
```
> $(REF Algebraic, std, variant)

Also didn't link

MetaLang commented:
Is there any reason you don't specialize `Option` on types that have a `null` value and use `null` in place of `defined`?

MetaLang commented on std/typecons.d:
```diff
@@ -6896,3 +6897,503 @@ unittest
+    }
+    else
+    {
+        template or(U)
```
is `inout` not usable here?

JakobOvrum replied:
Indeed it is not, that's why it's structured like it is.

JakobOvrum commented:
@MetaLang, this way supports `some(null)`, i.e. a non-empty null value, just like the Scala type. `Option` replaces null used to mean absence of an optional value, which doesn't cover every use of null.

JakobOvrum commented on std/typecons.d:
```diff
@@ -6896,3 +6897,503 @@ unittest
+                return isDefined ? typeof(return)(value) : typeof(return)(other);
+            }
+
+            Algebraic!(const T, U) or(lazy U other) const
```
`const` and `immutable` overloads are untested because `Algebraic` does not support non-mutable types...

MetaLang commented on std/typecons.d:
```diff
@@ -6896,3 +6897,503 @@ unittest
+ * Return the `front` or `back` of range when it is non-empty, or $(LREF none) otherwise.
+ * See_Also: $(LREF Option)
+ */
+Option!(ElementType!Range) frontOption(Range)(Range range)
```
Won't this be wrong in the case that Range is a string and its front returns `dchar`?

MetaLang replied:
Never mind, I see `ElementType` handles this.

andralex commented:
What is the relationship between this and `only`? At first sight it looks like a non-empty `Option` is `only(x)` and an empty `Option` is `only(T.init).drop`. If that's the case, we'd save a lot of trouble.

JakobOvrum commented:
Most of the work here is not duplicated by `only`. `Option!T` makes an effort to properly support any type of `T`, including non-mutable, non-copyable, non-default-constructable and nested types. Also, as `Option!T` is named and suitable for long-living variables (globals, TLS, member fields etc.), it destroys the held value when going *non-empty* → *empty*. Idioms for working with optional values like `getOrElse` and `frontOption` are also added and tested.

I think the best way would be to make unary `only` return `Option!T`. It is important that the type is named so it can be defined and initialized separately (typically in a [module] constructor) and show up in documentation when used as a return type. `only(a)` would be the same as `some(a)`, so the latter could be removed (and then it might make sense to move `Option` to std.range).

JakobOvrum commented:
Fixed compilation failure with latest DMD. `Option!T` is now only sliceable when `T` is copyable.

JakobOvrum commented:
Added `Option!V lookup(T : V[K], V, K)(T assocArray, auto ref K key)` as another fundamental artifact for working with option types. The name isn't particularly good, suggestions welcome.

ntrel commented:
I'm not convinced that `Option` should itself be a range. It makes code less clear that it's dealing with Options, and moreover `front` disguises the escape of a non-existent value. I know `empty` should be checked first for ranges, but when reading someone else's code you don't know if the range is just assumed to be non-empty.

I would much prefer all access to the value is null/assert-safe *or* uses some clear method like `assumeNonEmpty`. Option can still have a `range` method, but common things like map, filter probably should be added as Option methods.

> Added Option!V lookup(T : V[K], V, K)(T assocArray, auto ref K key) 

I think a version that returns an `Option!(V*)` is needed to replace `in`. Also, I don't think all functions producing an option need to go in std.typecons, `lookup` should go in std.array IMO.

JakobOvrum commented:
> I'm not convinced that Option should itself be a range. It makes code less clear that it's dealing with Options, and moreover front disguises the escape of a non-existent value. I know empty should be checked first for ranges, but when reading someone else's code you don't know if the range is just assumed to be non-empty.

Range code that assumes non-empty is wrong. Don't use `Option.front` directly either, it's just for passing to range algorithms.

Option being a range is the whole point. It's the same in Scala where it [has been very successful](https://tonymorris.github.io/blog/posts/scalaoption-cheat-sheet/).

> I think a version that returns an Option!(V*) is needed to replace in. Also, I don't think all functions producing an option need to go in std.typecons, lookup should go in std.array IMO.

How would `Option!(V*)` be used? If the contained pointer is nullable then the whole exercise is meaningless. The current `lookup` replaces `in`.

ntrel commented:
> How would Option!(V*) be used? If the contained pointer is nullable then the whole exercise is meaningless. The current lookup replaces in.

How do you modify `*(k in aa)` with your `lookup`? I'm suggesting to wrap `option(k in aa)`. If the result is not empty, the pointer, though nullable, would never be null unless modified by the caller. It could even return an Option of a wrapped pointer that can't be reassigned, guaranteeing the pointer is never null.

ntrel commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+
+    auto values = only(false, true, false).map!getX.map!option;
+    assert(values[0].empty);
+    assert(!values[1].empty && *values[1].front == 42);
```
> Don't use Option.front directly either, it's just for passing to range algorithms.

Please replace `front` here with something else (as this appears in the docs).

JakobOvrum replied:
Done. I ate my own dog food and rewrote it to use `equal`.

ntrel commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+    {
+        import core.exception : AssertError;
+        import std.exception : enforce;
+        enforce!AssertError(isDefined, T.stringof ~ " asssumption failed to hold");
```
typo: assumption

We could just use `getOrElse(T.init)` instead of this. Alternatively we could throw an `Exception` so the user can deliberately catch it and handle it.

JakobOvrum replied:
Fixed the typo.

It is an error in the source code of the program if this is called on an empty `Option`. The terminology follows Ali's lightning talk at DConf 2014 [slides](http://acehreli.org/AliCehreli_assumptions.pdf). It's reasonable to consider an interface that throws an exception, like `Option.get` in Scala, but my impression is that the Scala community would rather have had this instead, and they often use `get` as a (slightly lazy) shortcut for what this does.

If the caller wants `getOrElse(T.init)` then they should probably type just that.

tam4s commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+ */
+struct Option(T)
+{
+    private T value = T.init; // For when T has @disable this();
```
Can this support self referencing structs, like the ones below?
```
struct ListItem {
  int value;
  Option!ListItem next;
}
struct TreeNode {
  int value;
  Option!TreeNode left;
  Option!TreeNode right;
}
```

andralex replied:
@tam4s no, you'd need to use `Option` with pointers (or use classes).

tam4s replied:
If the value was stored in T[] or T* internally, then the above mentioned cases could work. I think storing it such way would be beneficial, at least for T types that has self-reference. ListItem & TreeNode with value semantics make sense; Currently I am having my own Option type for implementing that. Would be nice to use this one instead.

andralex commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+        }
+        else this.value = value;
+        this.isDefined = isDefined;
+    }
```
Not sure this function is sound. If `isDefined` is false, there's no need to do anything.

andralex commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+     */
+    void opAssign(NoneType)
+    {
+        if(isDefined)
```
space before `(` throughout

andralex commented on std/typecons.d:
```diff
@@ -7093,3 +7094,535 @@ unittest
+        {
+            import std.algorithm.mutation : swap;
+            this.isDefined = other.isDefined;
+            swap(this.value, other.value);
```
this is wasteful in a few cases

andralex commented:
>Most of the work here is not duplicated by only.

After making a pass through the definition and the documentation, I'm not convinced.

>Option!T makes an effort to properly support any type of T, including non-mutable, non-copyable, non-default-constructable and nested types.

That seems to mean `only` has bugs. Could those be fixed?

>Also, as Option!T is named and suitable for long-living variables (globals, TLS, member fields etc.), it destroys the held value when going non-empty → empty. Idioms for working with optional values such as getOrElse and frontOption are also added and tested.

Awesome. Again these seem bugs or shortcomings of `only`. Could we improve `only` to do that?

>I think the best way would be to make unary only return Option!T.

This is probably the right way to go, and by improving the existing definitions, not by adding a new name.

>It is important that the type is named so it can be defined and initialized separately (typically in a [module] constructor) and show up in documentation when used as a return type. only(a) would be the same as some(a), so the latter could be removed (and then it might make sense to move Option to std.range).

If `only` uses Voldemort, I'm okay with it receiving a legit name.

But please don't make the mistake of defining a almost-the-same-but-with-subtle-difference artifact just because you don't want to mess with the existing one.

JakobOvrum commented:
Yeah, you're right, this is too similar to unary `only` to separate them. I'll rework the PR on top of std.range and see how it turns out.

---

I noticed that `std.range.takeOne` is essentially `frontOption`, although we have no equivalent of `backOption`. From a clean slate I would have preferred `[front|back]Option`, but in the name of consistency, maybe `takeOne[Back]` wouldn't be the end of the world. This assumes that `[front|backOption]` carry their own weight anyway; it's probably better to keep the initial PR minimal and submit them as follow-ups, I've just been using this PR to brainstorm ideas about option types in D.

---

@ntrel, the current `lookup` does prevent double lookup but it doesn't provide the by-ref advantage of `in`. I was under the impression that `option(v)` dereferenced pointers but it doesn't and that's probably a good thing. Maybe `lookup` should return `Option!(V*)` for `V[K]`.

JakobOvrum commented:
* Moved to std.range
 * Removed `some` and changed unary `std.range.only` to return `Option!T`
 * Fixed @andralex's concerns about wasteful moves and swaps (and added space after `if` where it was missing)

---

About `takeOne` - it's a higher order range that supports accessing the front of the underlying range by reference, so it's not compatible with `frontOption`... it's an interesting difference.

---

About `lookup` - wrapping `option(key in aa)` is not useful, just write that as-is. What `lookup` provides is:
```d
auto aa = ["foo": 42];
assert(aa.lookup("foo").equal(only(42)));
// instead of
assert(option("foo" in aa).map!(p => *p).equal(only(42)));
```
This is the same inconvenience faced by any function that returns a null pointer to mean absence of value, such as `VariantN.peek`, so maybe a general solution for `T*` → `Option!T` is better.

---

No-one has said anything about it, but this PR still includes the arguably cargo-culty `NoneType` and `none` for explicit emptying. An alternative would be to just use default-initialization for empty construction, and `popFront` for emptying post-construction. I guess one issue is that `popFront` should not be called on an empty range while assigning to `none` is currently valid regardless of emptiness. Any thoughts?

---

I'm aware of the build failure, I'll investigate.

ntrel commented:
> NoneType 

I suppose `none` could be made a property so it doesn't have a named type.

> maybe a general solution for T* → Option!T is better

Yes, we could have:
```d
Option!T deref(T)(T*);
Option!T deref(T)(Option!(T*));
```
(I still think `lookup` should be in std.array, not sure what others think. Hopefully returning `Option` will become widespread across Phobos).

JakobOvrum commented:
Alright, here are the real differences between `Option!T` and `OnlyResult!(T, 1)`:
 * `Option!T` gives access to its held value by reference, which has many subtle implications, including supporting non-copyable types. `OnlyResult!(T, n)` gives access only by value, which is by design (the idea was to prevent `sort(only(3, 1, 2))` from compiling because it does not work as expected).
 * Because `OnlyResult!(T, 1)` also gives access by value, it can store the held value as `Unqual!T` while still being a range of `T` - but this relies on `is(Unqual!T : T)`, which can be false for user-defined types; e.g.`immutable(S)` where `struct S { int* p; }`. This seems like a design flaw.

Code in std.path relies on `OnlyResult!(immutable char, 1)` being reassignable (to itself), while `Option!(immutable char)` isn't. So far I haven't been able to reconcile this difference, but I'll give it another go later.

ntrel commented on std/range/package.d:
```diff
@@ -6767,6 +6767,543 @@ if (isForwardRange!Source && hasLength!Source)
+
+    auto values = only(false, true, false).map!getX.map!option;
+    assert(values[0].empty);
+    assert(values[1].map!(p => *p).equal(only(42)));
```
Could add an assert for `values[1].equal(only(getX(true)))` before this line. (And if we add `deref`, we can use it instead of map here and on line 7044).

ntrel commented:
> Because OnlyResult!(T, 1) also gives access by value, it can store the held value as Unqual!T while still being a range of T - but this relies on is(Unqual!T : T), which can be false for user-defined types; e.g.immutable(S) where struct S { int* p; }.

BTW I'm working on `Rebindable!(inout S)` which could solve these kind of problems. I think it's `@safe` and will hopefully submit it to Phobos. See https://github.com/ntrel/stuff/blob/master/rebindstruct.d.

wilzbach commented:
Merge failed - needs rebase :)

andralex commented:
rebase pls

wilzbach commented:
>> wilzbach commented on Apr 27
>> Merge failed - needs rebase :)
> rebase pls

Are u still alive @JakobOvrum? ;-)
Should we introduce a new category of "orphaned" PRs that are just lacking minor fixes and thus can be adopted by anyone?

andralex commented:
I see one way to push this forward: add a `none` function right after the existing `only` such that `none!int` returns an `Only!int` with nothing in it. This would cover the vast majority of uses of this PR.

JackStouffer commented:
No activity from PR author in nine months. Closing.

andralex commented:
Convert this to an enhancement request?

JackStouffer commented:
@andralex https://issues.dlang.org/show_bug.cgi?id=12679

s-ludwig commented:
@andralex 
> I see one way to push this forward: add a none function right after the existing `only` such that `none!int` returns an `Only!int` with nothing in it. This would cover the vast majority of uses of this PR.

Could you imagine renaming `Only!T` to one of `Option!T`/`Optional!T`/`Maybe!T` in the long term? It makes sense to go this way on a technical level, but seeing a declaration `void foo(Only!T t)`, one can hardly guess that this actually means that `t` is optional.

andralex commented:
Hmmm... neither name is very descriptive. The type really means "ZeroOrMore" whereas "Only" suggests exactly one and "Option" and "Optional" imply "zero or one". "Maybe" is arguably better. "Some" would be great but evokes "one or more" and promises to be a great source of confusion with Haskell's [some](https://hackage.haskell.org/package/base-4.9.1.0/docs/Control-Applicative.html#v:some). The same link suggests that "many" would be "zero or more".

"Bag" is a term I'd like - you have a "bag" of things, which may be empty, contain one item, or more items.

It's also a bummer that `only!int` is really zero ints...

s-ludwig commented:
Oh, I didn't realize that `only` can take multiple values! Then it doesn't really make sense of course. `alias Option(T) = OnlyResult!(T, 1);` could in theory work, although a proper type would be better in terms of documentation and generic code.
